storage/clickhousespanstore/reader_test.go

```diff
@@ -0,0 +1,1104 @@
+package clickhousespanstore
+
+import (
+	"context"
+	"database/sql/driver"
+	"encoding/json"
+	"fmt"
+	"math"
+	"math/rand"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/DATA-DOG/go-sqlmock"
+	"github.com/gogo/protobuf/proto"
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	testOperationsTable = "test_operations_table"
+	testNumTraces       = 10
+	testSpansInTrace    = 2
+	testOperationCount  = 10
+)
+
+var testStartTime = time.Date(2010, 3, 15, 7, 40, 0, 0, time.UTC)
+
+func TestTraceReader_FindTraceIDs(t *testing.T) {
+	db, mock, err := getDbMock()
+	require.NoError(t, err, "an error was not expected when opening a stub database connection")
+	defer db.Close()
+
+	traceReader := NewTraceReader(db, testOperationsTable, testIndexTable, testSpansTable)
+	service := "service"
+	start := testStartTime
+	end := start.Add(24 * time.Hour)
+	fullDuration := end.Sub(start)
+	duration := fullDuration
+	for i := 0; i < maxProgressiveSteps; i++ {
+		duration /= 2
+	}
+	params := spanstore.TraceQueryParameters{
+		ServiceName:  service,
+		NumTraces:    testNumTraces,
+		StartTimeMin: start,
+		StartTimeMax: end,
+	}
+
+	expectedTraceIDs := make([]model.TraceID, testNumTraces)
+	traceIDValues := make([]driver.Value, testNumTraces)
+	for i := range expectedTraceIDs {
+		traceID := model.TraceID{Low: uint64(i)}
+		expectedTraceIDs[i] = traceID
+		traceIDValues[i] = traceID.String()
+	}
+
+	found := traceIDValues[:0]
+	endArg := end
+	for i := 0; i < maxProgressiveSteps; i++ {
```

Comment:
could you please comment what this code block does? Why the for loop is needed and what `maxProgressiveSteps` mean?

Reply:
It's needed to generate timestamps used as borders in TraceReader queries. There, start-end interval is halved maxProgressiveSteps times. Then left border is gotten for query subtracting that interval from the end. After that, interval is increased twice its size.